### PR TITLE
improvement: smaller block status map during IBD

### PR DESCRIPTION
**Pull request by yangby-cryptape**
Not conflict with #2609:

- This PR to make block status map to be smaller.

  After this PR, the max size of block status map during IBD: `4*10^6 -> 5*10^4` (current mainnet has about `4*10^6` blocks, and **I only test once**).

- #2609 to cache block status map into disk if the size of block status map is bigger than a threshold (default: `10^6`, `1/4` of current mainchain).

  If this PR merged, #2609 will not work in almost all scenarios at present.

  But I don't close it because:
  - The max size of block status map will be increased if network is poor or too many forked blocks.
  - We can set it to a smaller value, #2609 provide an option.
  - That PR refactor the `HeaderMap` to `CachedHashMap`, so we can reuse the temporary storage for other `HashMap`s later.

Both of these two PRs could be merged.

### CODE REVIEW

**Comment by yangby-cryptape** [comment]
sync-test

**Comment by nervos-bot-user** [comment]
#### Sync Test Result
  - CKB Version: a251023f546ac67b2ba0ac8405645d683418736e
  - Instance Region: eu-west-2
  - Sync Start: 2021-03-12 00:15:32
  - Sync End: 2021-03-12 02:23:13
  - Sync Height: 3938449
  - Sync Speed: 514 blocks/s

File: sync/src/types/mod.rs [hunk 1]
```diff
@@ -1984,15 +1982,19 @@ impl ActiveChain {
...
-                let verified = self
-                    .snapshot
-                    .get_block_ext(block_hash)
-                    .map(|block_ext| block_ext.verified);
-                match verified {
-                    None => BlockStatus::UNKNOWN,
-                    Some(None) => BlockStatus::BLOCK_STORED,
-                    Some(Some(true)) => BlockStatus::BLOCK_VALID,
-                    Some(Some(false)) => BlockStatus::BLOCK_INVALID,
+                if self.state.header_map.contains_key(block_hash) {
```
Comment (driftluo) [inline sync/src/types/mod.rs:1985]: It contains BLOCK_INVALID status inside, so you can’t write it like this

**Review by driftluo — APPROVED** [review]

**Review by keroro520 — APPROVED** [review]

**Review by quake — APPROVED** [review]

**Comment by driftluo** [comment]
bors r=driftluo,keroro520,quake

**Pull request merged**